### PR TITLE
Implement pdf functions and add type 0, 2 and 4 function tests

### PR DESCRIPTION
**Pull request by BobLd**
Implement pdf functions and add type 0, 2 and 4 function tests

<del>I could not find any document with function type 3 example, so no tests so far (PDFBOX only tests type 4). Test for type 0 are basic but could not find any output example. I expect we will be able to test better when rendering to image is implemented </del>

EDIT: I found tests for actual values of type 0 function, now added and fixed the function 0 GetSamples(). bits had to be itereated starting from the end

EDIT 2: I tested function type 3 rendering to image (mainly used in shading it seems)

### CODE REVIEW

**Comment by BobLd** [comment]
Should be good to go, please let me know if any comment

**Comment by BobLd** [comment]
Good to go now, function type 0 are well tested

**Comment by BobLd** [comment]
Please hold before reviewing/merging, I found some errors with Function type 3, this will need some refactoring

**Comment by BobLd** [comment]
Type 3 are now fixed and I cleaned to csproj stuff. It's finally good to go. Let me know if any question

**Comment by BobLd** [comment]
Fixed type 4 function stack

**Comment by EliotJones** [comment]
Thanks @BobLd I'm in the process of moving countries again for like the 5th time in 5 months but will try to get to this soon

**Review by EliotJones — APPROVED** [review]
Thanks for all the hard work, I haven't reviewed for logic just eyeballing it but the tests are a nice touch :+1:

File: src/UglyToad.PdfPig/Functions/PdfFunctionType0.cs
Comment (BobLd) [reply src/UglyToad.PdfPig/Functions/PdfFunctionType0.cs:351]: okay thanks for letting me know, wasn't aware. I'll remove that in my next PR

**Pull request merged**
